Sources/MapboxCoreNavigation/BillingHandler.swift

```diff
@@ -0,0 +1,400 @@
+        return .init(service: service)
+    }()
+
+    /// The billing service which is used to send billing events.
+    private let billingService: BillingService
+
+    /**
+     A lock which serializes access to variables with underscore: `_sessions` etc.
+     As a convention, all class-level identifiers that starts with `_` should be executed with locked `lock`.
+     */
+    private let lock: NSLock = .init()
+
+    /// All currently active sessions. Running or paused. When session is stopped, it is removed from this variable.
+    private var _sessions: [UUID: Session] = [:]
+
+    private let logger: OSLog = .init(subsystem: "com.mapbox.navigation", category: "Billing")
+
+    /**
+     The state of the billing session.
+
+     - important: This variable is safe to use from any thread.
+     - parameter uuid: Session UUID which is provided in `BillingHandler.beginBillingSession(for:uuid:)`.
+     */
+    func sessionState(uuid: UUID) -> SessionState {
+        lock.lock(); defer {
+            lock.unlock()
+        }
+
+        guard let session = _sessions[uuid] else {
+            return .stopped
+        }
+
+        if session.isPaused {
+            return .paused
+        }
+        else {
+            return .running
+        }
+    }
+
+    /// The token to use for service requests like `Directions` etc. 
+    var serviceSkuToken: String {
+        let sessionTypes: [BillingHandler.SessionType] = [.activeGuidance, .freeDrive]
+
+        for sessionType in sessionTypes {
+            switch billingService.getSessionStatus(for: sessionType) {
+            case .running:
+                return billingService.getSKUTokenIfValid(for: sessionType)
+            case .paused, .stopped:
+                continue
+            }
+        }
+
+        return ""
+    }
+
+    private init(service: BillingService) {
+        self.billingService = service
+    }
+
+    /**
+     Starts a new billing session of the given `sessionType` identified by `uuid`.
+
+     The `uuid` that is used to create a billing session must be provided in the following methods to perform
+     relevant changes to the started billing session:
+     - `BillingHandler.stopBillingSession(with:)`
+     - `BillingHandler.pauseBillingSession(with:)`
+     - `BillingHandler.resumeBillingSession(with:)`
+
+     - Parameters:
+     - sessionType: The type of the billing session.
+     - uuid: The unique identifier of the billing session.
+     */
+    func beginBillingSession(for sessionType: SessionType, uuid: UUID) {
+        lock.lock()
+
+        if var existingSession = _sessions[uuid] {
+            existingSession.isPaused = false
+            _sessions[uuid] = existingSession
+        }
+        else {
+            let session = Session(type: sessionType, isPaused: false)
+            _sessions[uuid] = session
+        }
+
+        let triggerBeginSessionEvents = billingService.getSessionStatus(for: sessionType) == .stopped
```

Comment:
Maybe it's better to use `switch` here to prevent call duplication? Or at least store the result of the call in a local variable.

Reply:
Thanks, I will try to improve this code.